File: patterns/_tests/BehavioralPatterns/Interpreter/cpp/unit.test.cpp
// ============================================================
//  Unit Tests — Interpreter: Arithmetic Expression Evaluator (C++)
//
//  Compile: g++ -std=c++17 -DTESTING -o unit_test unit.test.cpp \
//               ../../../../BehavioralPatterns/Interpreter/cpp/cpp.cpp
//  Run: ./unit_test
// ============================================================

#define TESTING
#include "../../../../BehavioralPatterns/Interpreter/cpp/cpp.cpp"
#include <cassert>
#include <iostream>
#include <stdexcept>
#include <string>

static int passed = 0;
static int failed = 0;

#define TEST(name, body) \
    do { \
        try { body; ++passed; std::cout << "  PASS  " << name << "\n"; } \
        catch (const std::exception& e) { ++failed; std::cout << "  FAIL  " << name << " : " << e.what() << "\n"; } \
        catch (...) { ++failed; std::cout << "  FAIL  " << name << " : unknown exception\n"; } \
    } while (false)

#define ASSERT_EQ(a, b)   do { if (!((a) == (b))) throw std::runtime_error("Expected " #a " == " #b); } while(false)
#define ASSERT_NEAR(a, b) do { if (std::abs((a) - (b)) > 0.001) throw std::runtime_error("Values not close: " #a " vs " #b); } while(false)
#define ASSERT_THROWS(expr, ExType) \
    do { bool threw = false; try { expr; } catch (const ExType&) { threw = true; } \
         if (!threw) throw std::runtime_error("Expected exception " #ExType " not thrown"); } while(false)


int main() {
    std::cout << "\n  Unit Tests — Interpreter (C++)\n";
    std::cout << "  " << std::string(50, '-') << "\n\n";

    // ── Context ───────────────────────────────────────────────
    TEST("Context: stores and retrieves variable", {
        Context ctx; ctx.set("x", 42);
        ASSERT_NEAR(ctx.get("x"), 42.0);
    });

    TEST("Context: throws on missing variable", {
        Context ctx;
        ASSERT_THROWS(ctx.get("missing"), std::runtime_error);
    });

    TEST("Context: overwrites existing variable", {
        Context ctx; ctx.set("x", 1); ctx.set("x", 99);
        ASSERT_NEAR(ctx.get("x"), 99.0);
    });

    // ── NumberExpression ──────────────────────────────────────
    TEST("NumberExpression: returns literal value", {
        Context ctx;
        ASSERT_NEAR(num(7)->interpret(ctx), 7.0);
    });

    TEST("NumberExpression: ignores context", {
        Context ctx; ctx.set("x", 999);
        ASSERT_NEAR(num(3)->interpret(ctx), 3.0);
    });

    TEST("NumberExpression: display renders integer", {
        ASSERT_EQ(num(5)->display(), std::string("5"));
    });

    // ── VariableExpression ────────────────────────────────────
    TEST("VariableExpression: resolves from context", {
        Context ctx; ctx.set("a", 10);
        ASSERT_NEAR(var("a")->interpret(ctx), 10.0);
    });

    TEST("VariableExpression: throws when not in context", {
        Context ctx;
        ASSERT_THROWS(var("z")->interpret(ctx), std::runtime_error);
    });

    TEST("VariableExpression: display returns name", {
        ASSERT_EQ(var("a")->display(), std::string("a"));
    });

    // ── AddExpression ─────────────────────────────────────────
    TEST("AddExpression: adds two numbers", {
        Context ctx;
        ASSERT_NEAR(add(num(3), num(4))->interpret(ctx), 7.0);
    });

    TEST("AddExpression: adds two variables", {
        Context ctx; ctx.set("a", 10); ctx.set("b", 5);
        ASSERT_NEAR(add(var("a"), var("b"))->interpret(ctx), 15.0);
    });

    // ── SubtractExpression ────────────────────────────────────
    TEST("SubtractExpression: subtracts right from left", {
        Context ctx;
        ASSERT_NEAR(sub(num(10), num(3))->interpret(ctx), 7.0);
    });

    TEST("SubtractExpression: result can be negative", {
        Context ctx;
        ASSERT_NEAR(sub(num(2), num(5))->interpret(ctx), -3.0);
    });

    // ── MultiplyExpression ────────────────────────────────────
    TEST("MultiplyExpression: multiplies two numbers", {
        Context ctx;
        ASSERT_NEAR(mul(num(4), num(5))->interpret(ctx), 20.0);
    });

    TEST("MultiplyExpression: multiplies variable by number", {
        Context ctx; ctx.set("a", 6);
        ASSERT_NEAR(mul(var("a"), num(7))->interpret(ctx), 42.0);
    });

    // ── DivideExpression ──────────────────────────────────────
    TEST("DivideExpression: divides left by right", {
        Context ctx;
        ASSERT_NEAR(div_(num(10), num(2))->interpret(ctx), 5.0);
    });

    TEST("DivideExpression: throws on zero", {
        Context ctx;
        ASSERT_THROWS(div_(num(5), num(0))->interpret(ctx), std::runtime_error);
    });

    TEST("DivideExpression: produces fractional result", {
        Context ctx;
        ASSERT_NEAR(div_(num(10), num(4))->interpret(ctx), 2.5);
    });

    // ── Composite trees ───────────────────────────────────────
    TEST("Composite: diff of squares (a+b)*(a-b) = 75", {
        Context ctx; ctx.set("a", 10); ctx.set("b", 5);
        auto expr = mul(add(var("a"), var("b")), sub(var("a"), var("b")));
        ASSERT_NEAR(expr->interpret(ctx), 75.0);
    });

    TEST("Composite: same tree, different contexts", {
        auto a = var("a"); auto b = var("b");
        auto expr = mul(add(a, b), sub(a, b));
        Context c1; c1.set("a", 10); c1.set("b", 5);
        Context c2; c2.set("a",  6); c2.set("b", 2);
        ASSERT_NEAR(expr->interpret(c1), 75.0);
        ASSERT_NEAR(expr->interpret(c2), 32.0);
    });

    TEST("Composite: (3+4)*2 = 14", {
        Context ctx;
        ASSERT_NEAR(mul(add(num(3), num(4)), num(2))->interpret(ctx), 14.0);
    });

    // ── Summary ───────────────────────────────────────────────
    std::cout << "\n  " << std::string(50, '-') << "\n";
    std::cout << "  " << passed << " passed, " << failed << " failed\n\n";
    return failed > 0 ? 1 : 0;
}

File: patterns/_tests/BehavioralPatterns/Interpreter/typescript/unit.test.ts
// ============================================================
//  Unit Tests — Interpreter: Arithmetic Expression Evaluator (TypeScript)
//
//  Run: npx mocha --require ts-node/register unit.test.ts
// ============================================================

import { describe, it } from 'mocha';
import assert from 'assert';
import {
  Context,
  NumberExpression, VariableExpression,
  AddExpression, SubtractExpression, MultiplyExpression, DivideExpression,
  FloorDivideExpression, ModuloExpression, PowerExpression,
  NegateExpression, AbsExpression, SqrtExpression, InverseExpression,
  FactorialExpression, PowerOfTenExpression, LogExpression,
} from '../../../../BehavioralPatterns/Interpreter/typescript/typescript';

const ctx0 = new Context();

// ── Context ───────────────────────────────────────────────────
describe('Context', () => {
  it('stores and retrieves a variable', () => {
    assert.strictEqual(new Context().set('x', 42).get('x'), 42);
  });
  it('throws on undefined variable', () => {
    assert.throws(() => new Context().get('missing'), /Undefined variable/);
  });
  it('overwrites an existing variable', () => {
    assert.strictEqual(new Context().set('x', 1).set('x', 99).get('x'), 99);
  });
  it('set() returns context for chaining', () => {
    const ctx = new Context(); assert.strictEqual(ctx.set('a', 1), ctx);
  });
});


// ── Terminal expressions ──────────────────────────────────────
describe('NumberExpression', () => {
  it('returns its literal value', () => assert.strictEqual(new NumberExpression(7).interpret(ctx0), 7));
  it('ignores the context',       () => assert.strictEqual(new NumberExpression(3).interpret(new Context().set('x', 99)), 3));
  it('toString renders the number', () => assert.strictEqual(new NumberExpression(5).toString(), '5'));
});

describe('VariableExpression', () => {
  it('resolves from context',          () => assert.strictEqual(new VariableExpression('a').interpret(new Context().set('a', 10)), 10));
  it('throws when missing',            () => assert.throws(() => new VariableExpression('z').interpret(ctx0), /Undefined variable/));
  it('toString returns variable name', () => assert.strictEqual(new VariableExpression('a').toString(), 'a'));
});


// ── Binary operators ──────────────────────────────────────────
describe('AddExpression', () => {
  it('adds two numbers',   () => assert.strictEqual(new AddExpression(new NumberExpression(3), new NumberExpression(4)).interpret(ctx0), 7));
  it('display wraps in parens', () => assert.ok(new AddExpression(new NumberExpression(1), new NumberExpression(2)).toString().startsWith('(')));
});

describe('SubtractExpression', () => {
  it('subtracts right from left', () => assert.strictEqual(new SubtractExpression(new NumberExpression(10), new NumberExpression(3)).interpret(ctx0), 7));
  it('result can be negative',    () => assert.strictEqual(new SubtractExpression(new NumberExpression(2), new NumberExpression(5)).interpret(ctx0), -3));
});

describe('MultiplyExpression', () => {
  it('multiplies two numbers', () => assert.strictEqual(new MultiplyExpression(new NumberExpression(4), new NumberExpression(5)).interpret(ctx0), 20));
});

describe('DivideExpression', () => {
  it('divides left by right',      () => assert.strictEqual(new DivideExpression(new NumberExpression(10), new NumberExpression(2)).interpret(ctx0), 5));
  it('throws on division by zero', () => assert.throws(() => new DivideExpression(new NumberExpression(5), new NumberExpression(0)).interpret(ctx0), /Division by zero/));
  it('produces fractional result', () => assert.strictEqual(new DivideExpression(new NumberExpression(10), new NumberExpression(4)).interpret(ctx0), 2.5));
});

describe('FloorDivideExpression', () => {
  it('floors the quotient',        () => assert.strictEqual(new FloorDivideExpression(new NumberExpression(17), new NumberExpression(5)).interpret(ctx0), 3));
  it('throws on division by zero', () => assert.throws(() => new FloorDivideExpression(new NumberExpression(5), new NumberExpression(0)).interpret(ctx0), /Division by zero/));
  it('display contains //',        () => assert.ok(new FloorDivideExpression(new NumberExpression(1), new NumberExpression(2)).toString().includes('//')));
});

describe('ModuloExpression', () => {
  it('returns the remainder',      () => assert.strictEqual(new ModuloExpression(new NumberExpression(17), new NumberExpression(5)).interpret(ctx0), 2));
  it('throws on division by zero', () => assert.throws(() => new ModuloExpression(new NumberExpression(5), new NumberExpression(0)).interpret(ctx0), /Division by zero/));
  it('display contains %',         () => assert.ok(new ModuloExpression(new NumberExpression(1), new NumberExpression(2)).toString().includes('%')));
});

describe('PowerExpression', () => {
  it('raises base to exponent',  () => assert.strictEqual(new PowerExpression(new NumberExpression(2), new NumberExpression(10)).interpret(ctx0), 1024));
  it('handles fractional power', () => assert.ok(Math.abs(new PowerExpression(new NumberExpression(4), new NumberExpression(0.5)).interpret(ctx0) - 2) < 0.001));
  it('display contains **',      () => assert.ok(new PowerExpression(new NumberExpression(2), new NumberExpression(3)).toString().includes('**')));
});


// ── Unary functions ───────────────────────────────────────────
describe('NegateExpression', () => {
  it('negates positive', () => assert.strictEqual(new NegateExpression(new NumberExpression(5)).interpret(ctx0), -5));
  it('negates negative', () => assert.strictEqual(new NegateExpression(new NumberExpression(-3)).interpret(ctx0), 3));
});

describe('AbsExpression', () => {
  it('returns absolute value of negative', () => assert.strictEqual(new AbsExpression(new NumberExpression(-9)).interpret(ctx0), 9));
  it('leaves positive unchanged',          () => assert.strictEqual(new AbsExpression(new NumberExpression(7)).interpret(ctx0), 7));
});

describe('SqrtExpression', () => {
  it('returns correct square root',    () => assert.strictEqual(new SqrtExpression(new NumberExpression(100)).interpret(ctx0), 10));
  it('throws on negative input',       () => assert.throws(() => new SqrtExpression(new NumberExpression(-1)).interpret(ctx0), /negative/));
});

describe('InverseExpression', () => {
  it('returns 1/x',               () => assert.strictEqual(new InverseExpression(new NumberExpression(4)).interpret(ctx0), 0.25));
  it('throws on zero',            () => assert.throws(() => new InverseExpression(new NumberExpression(0)).interpret(ctx0), /Division by zero/));
});

describe('FactorialExpression', () => {
  it('computes 0! = 1',  () => assert.strictEqual(new FactorialExpression(new NumberExpression(0)).interpret(ctx0), 1));
  it('computes 5! = 120',() => assert.strictEqual(new FactorialExpression(new NumberExpression(5)).interpret(ctx0), 120));
  it('computes 6! = 720',() => assert.strictEqual(new FactorialExpression(new NumberExpression(6)).interpret(ctx0), 720));
  it('throws on negative',() => assert.throws(() => new FactorialExpression(new NumberExpression(-1)).interpret(ctx0), /negative/));
});

describe('PowerOfTenExpression', () => {
  it('10^2 = 100',  () => assert.strictEqual(new PowerOfTenExpression(new NumberExpression(2)).interpret(ctx0), 100));
  it('10^0 = 1',    () => assert.strictEqual(new PowerOfTenExpression(new NumberExpression(0)).interpret(ctx0), 1));
});

describe('LogExpression', () => {
  it('log(100) = 2',  () => assert.strictEqual(new LogExpression(new NumberExpression(100)).interpret(ctx0), 2));
  it('log(1) = 0',    () => assert.strictEqual(new LogExpression(new NumberExpression(1)).interpret(ctx0), 0));
  it('throws on zero',() => assert.throws(() => new LogExpression(new NumberExpression(0)).interpret(ctx0), /non-positive/));
  it('throws on negative',() => assert.throws(() => new LogExpression(new NumberExpression(-5)).interpret(ctx0), /non-positive/));
});


// ── Composite trees ───────────────────────────────────────────
describe('Composite expression trees', () => {
  it('(a + b) * (a - b) equals a² - b²', () => {
    const a = new VariableExpression('a'), b = new VariableExpression('b');
    const ctx = new Context().set('a', 10).set('b', 5);
    assert.strictEqual(new MultiplyExpression(new AddExpression(a, b), new SubtractExpression(a, b)).interpret(ctx), 75);
  });

  it('sqrt(abs(a - b)) composes correctly', () => {
    const a = new VariableExpression('a'), b = new VariableExpression('b');
    const ctx = new Context().set('a', 7).set('b', 23);
    const expr = new SqrtExpression(new AbsExpression(new SubtractExpression(a, b)));
    assert.strictEqual(expr.interpret(ctx), 4);
  });

  it('same tree, different contexts', () => {
    const a = new VariableExpression('a'), b = new VariableExpression('b');
    const expr = new MultiplyExpression(new AddExpression(a, b), new SubtractExpression(a, b));
    assert.strictEqual(expr.interpret(new Context().set('a', 10).set('b', 5)), 75);
    assert.strictEqual(expr.interpret(new Context().set('a', 6).set('b', 2)), 32);
  });

  it('10^abs(b) composes PowerOfTen with Abs', () => {
    const b = new VariableExpression('b');
    const ctx = new Context().set('b', -3);
    assert.strictEqual(new PowerOfTenExpression(new AbsExpression(b)).interpret(ctx), 1000);
  });
});
